server/src/main/java/com/microsoft/java/bs/core/internal/services/LifecycleService.java

```diff
@@ -119,61 +134,117 @@ public void exit() {
+    File gradleJavaHome;
+    if (StringUtils.isNotEmpty(filePath)) {
+      gradleJavaHome = new File(filePath);
+    } else {
+      gradleJavaHome = connector.getGradleJavaHome(rootUri);
     }
 
-    if (StringUtils.isBlank(preferences.getGradleJavaHome())) {
-      String gradleVersion = "";
-      GradleBuildKind buildKind = Utils.getEffectiveBuildKind(new File(rootUri), preferences);
-      Map<String, String> map = TelemetryUtils.getMetadataMap("buildKind", buildKind.name());
-      LOGGER.log(Level.INFO, "Use build kind: " + buildKind.name(), map);
-      if (buildKind == GradleBuildKind.SPECIFIED_VERSION) {
-        gradleVersion = preferences.getGradleVersion();
-      } else {
-        gradleVersion = connector.getGradleVersion(rootUri);
-      }
+    GradleBuildKind buildKind = Utils.getEffectiveBuildKind(new File(rootUri), preferences);
+    Map<String, String> map = TelemetryUtils.getMetadataMap("buildKind", buildKind.name());
+    LOGGER.log(Level.INFO, "Use build kind: " + buildKind.name(), map);
+
+    String gradleVersion;
+    if (buildKind == GradleBuildKind.SPECIFIED_VERSION) {
+      gradleVersion = preferences.getGradleVersion();
+    } else {
+      gradleVersion = connector.getGradleVersion(rootUri);
+    }
+
+    if (StringUtils.isEmpty(gradleVersion)) {
+      return null;
+    }
+
+    map = TelemetryUtils.getMetadataMap("gradleVersion", gradleVersion);
+    LOGGER.log(Level.INFO, "Gradle version: " + gradleVersion, map);
+
+    String latestCompatibleVersion = Utils.getLatestCompatibleJavaVersion(gradleVersion);
+    String oldestCompatibleVersion = Utils.getOldestCompatibleJavaVersion();
+
+    if (StringUtils.isNotBlank(latestCompatibleVersion)) {
 
-      if (StringUtils.isNotBlank(gradleVersion)) {
-        map = TelemetryUtils.getMetadataMap("gradleVersion", gradleVersion);
-        LOGGER.log(Level.INFO, "Gradle version: " + gradleVersion, map);
-        String highestJavaVersion = Utils.getHighestCompatibleJavaVersion(gradleVersion);
-        File jdkInstallation = getJdkToLaunchDaemon(preferences.getJdks(), highestJavaVersion);
-        if (jdkInstallation != null) {
-          preferences.setGradleJavaHome(jdkInstallation.getAbsolutePath());
+      // Use GradleJavaHome if compatible
+      if (gradleJavaHome != null) {
+
+        try {
+          String gradleJavaHomeVersion = JavaUtils.getJavaVersionFromFile(gradleJavaHome);
+          if (
+              JavaUtils.isCompatible(
+                  gradleJavaHomeVersion,
+                  oldestCompatibleVersion,
+                  latestCompatibleVersion
+              )
+          ) {
+            return gradleJavaHome;
+          }
+        } catch (IOException e) {
+          LOGGER.severe("Invalid GradleJavaHome: " + e.getMessage());
         }
+
+      }
+
+      // Pick a compatible JDK from the JDKs available in Preferences
+      if (preferences.getJdks() != null && preferences.getJdks().isEmpty()) {
+        return getJdkToLaunchDaemon(
```

Comment:
If the preference has set some jdks but no one is compatible with gradle, since we return here, the client will not be notified.

Reply:
Addressed in commit [c7f5d83](https://github.com/microsoft/build-server-for-gradle/pull/165/commits/c7f5d838683fa585efa8d5ffb25ebe9b87d112ea)